pyccel/stdlib/cwrapper/cwrapper.c

```diff
@@ -0,0 +1,273 @@
+{
+	float	f;
+	double	out;
+
+	out = PyFloat_AsDouble(o);
+
+	f = (float)out;
+
+	return	f;
+}
+
+double	PyDouble_to_Double(PyObject *o)
+{
+	double	d;
+
+	d = PyFloat_AsDouble(o);
+
+	return	d;
+}
+
+
+/*
+ * Functions : Cast functions
+ * ---------------------------
+ * Some of the function used below are based on C/python api
+ * with more tolerance to different precisions and complex type.
+ *	Parameterss	:
+ *		C object
+ *	Returns     :
+ *		o  : python object
+ * reference of the used c/python api function
+ * ---------------------------------------------------
+ * https://docs.python.org/3/c-api/complex.html#c.PyComplex_FromDoubles
+ * https://docs.python.org/3/c-api/float.html#c.PyFloat_FromDouble
+ * https://docs.python.org/3/c-api/long.html#c.PyLong_FromLongLong
+ */
+
+PyObject	*Complex_to_PyComplex(double complex c)
+{
+	double		real_part;
+	double		imag_part;
+	PyObject	*o;
+
+	real_part = creal(c);
+	imag_part = cimag(c);
+	o = PyComplex_FromDoubles(real_part, imag_part);
+
+	return o;
+}
+
+PyObject	*Bool_to_PyBool(bool b)
+{
+	return b == true ? Py_True : Py_False;
+}
+
+PyObject	*Int_to_PyLong(int64_t i)
+{
+	PyObject	*o;
+
+	o = PyLong_FromLongLong((long long) i);
+
+	return o;
+}
+
+PyObject	*Double_to_PyDouble(double d)
+{
+	PyObject	*o;
+
+	o = PyFloat_FromDouble(d);
+
+	return o;
+}
+
+/*
+ * Function: PyArray_to_ndarray
+ * ----------------------------
+ * A Cast function that convert numpy array variable into ndarray variable,
+ * by copying its information and data to a new variable of type ndarray struct
+ * and return this variable to be used inside c code.
+ * 	Parameters	:
+ *		o 	  : python array object
+ * 	Returns		:
+ *		array : c ndarray
+ *
+ * reference of the used c/numpy api function
+ * -------------------------------------------
+ * https://numpy.org/doc/stable/reference/c-api/array.html
+ */
+
+
+t_ndarray		PyArray_to_ndarray(PyArrayObject *o)
+{
+	t_ndarray	array;
+
+	array.nd          = PyArray_NDIM(o);
+	array.raw_data    = PyArray_DATA(o);
+	array.type_size   = PyArray_ITEMSIZE(o);
+	array.type        = PyArray_TYPE(o);
+	array.length      = PyArray_SIZE(o);
+	array.buffer_size = PyArray_NBYTES(o);
+	array.shape       = numpy_to_ndarray_shape(PyArray_SHAPE(o), array.nd);
+	array.strides     = numpy_to_ndarray_strides(PyArray_STRIDES(o), array.type_size, array.nd);
+	array.is_view     = 1;
+
+	return array;
+}
+
+
+/*
+ * Function: Check_Array
+ * --------------------
+ * Check Python Object (ArrayType, Rank, Order):
+ *
+ *  Parameters :
+ * 		a     : Python Object
+ *  	rank  : The desired rank
+ *  	order : The desired order
+ *               0 if the check order is not needed
+ *
+ *  Returns    :
+ * 		reference to PyArray Object
+ *      returns NULL on error
+ * reference of the used c/python api function
+ * -------------------------------------------
+ * https://numpy.org/doc/stable/reference/c-api/array.html
+ */
+
+PyArrayObject	*Check_Array(PyObject *a, int rank, int flags)
```

Comment:
Shouldn't `Check_Array` also check the dtype?

Reply:
Is this done separately to avoid having lots of if blocks for the correct error message?
I'm not sure whether to suggest passing the string as an argument, or if that will lead to a big ugly function call

Reply:
using interface and to avoid multiple type check the check data is only written when needed in the converter function

Reply:
ok 👍

Reply:
==> I'm not sure whether to suggest passing the string as an argument, or if that will lead to a big ugly function call ?
i was thinking about something like  global static array so when check type is needed i can pass the type num as arguments or 0 if not needed   , and use array for error code